File: libsrc/feassemble/IntegratorExplicit.cc
// -*- C++ -*-
//
// ======================================================================
//
//                           Brad T. Aagaard
//                        U.S. Geological Survey
//
// {LicenseText}
//
// ======================================================================
//

#include <portinfo>

#include "IntegratorExplicit.hh" // implementation of class methods

#include <assert.h> // USES assert()

// ----------------------------------------------------------------------
// Constructor
pylith::feassemble::IntegratorExplicit::IntegratorExplicit(void) :
  Integrator(),
  _dt(0.0),
  _dtm1(0.0)
{ // constructor
} // constructor

// ----------------------------------------------------------------------
// Destructor
pylith::feassemble::IntegratorExplicit::~IntegratorExplicit(void)
{ // destructor
} // destructor
  
// ----------------------------------------------------------------------
// Copy constructor
pylith::feassemble::IntegratorExplicit::IntegratorExplicit(const IntegratorExplicit& i) :
  Integrator(i),
  _dt(i._dt),
  _dtm1(i._dtm1)
{ // copy constructor
} // copy constructor

// ----------------------------------------------------------------------
// Set time step for advancing from time t to time t+dt.
void
pylith::feassemble::IntegratorExplicit::timeStep(const double dt)
{ // timeStep
  _dtm1 = _dt;
  _dt = dt;
  assert(_dt == _dtm1); // For now, don't allow variable time step
} // timeStep

// ----------------------------------------------------------------------
// Get stable time step for advancing from time t to time t+dt.
double
pylith::feassemble::IntegratorExplicit::stableTimeStep(void) const
{ // stableTimeStep
  // Default is current time step
  return _dt;
} // stableTimeStep


// End of file 

File: pylith/feassemble/IntegratorExplicit.py
#!/usr/bin/env python
#
# ----------------------------------------------------------------------
#
#                           Brad T. Aagaard
#                        U.S. Geological Survey
#
# <LicenseText>
#
# ----------------------------------------------------------------------
#

## @file pylith/feassemble/IntegratorExplicit.py
##
## @brief Python object for explicit time integration of actions with
## finite-elements.
##
## Factory: integrator

from Integrator import Integrator

# IntegratorInertia class
class IntegratorExplicit(Integrator):
  """
  Python object for explicit integration of operator actions with
  finite-elements.

  Factory: integrator.
  """

  # PUBLIC METHODS /////////////////////////////////////////////////////

  def __init__(self, name="integratorexplicit"):
    """
    Constructor.
    """
    Integrator.__init__(self, name)
    return


  def timeStep(self, t):
    """
    Set time step for advancing from time t to time t+dt.
    """
    self.cppHandle.timeStep(t.value)
    return


  def stableTimeStep(self):
    """
    Get stable time step for advancing from time t to time t+dt.
    """
    return self.cppHandle.getStableTimeStep()


  def integrateConstant(self, fieldOut, fieldInT, fieldInTmdt):
    """
    Integrate constant term for dynamic terms for finite-elements.
    """
    self.cppHandle.integrateConstant(fieldOut, fieldInT, fieldInTmdt,
                                     self.mesh.cppHandle)
    return


  def integrateJacobian(self, jacobian, fieldInT, coords):
    """
    Integrate Jacobian term for dynamic terms for finite-elements.
    """
    self.cppHandle.integrateJacobian(jacobian, fieldInT, self.mesh.cppHandle)
    return


# End of file 
